From 7556111f007c98f11adfa27c492d73b775886d9d Mon Sep 17 00:00:00 2001
From: Jeremy Harris <jgh146exb@wizmail.org>
Date: Thu, 21 Mar 2019 20:01:03 +0000
Subject: [PATCH 13/29] Harden plaintext authenticator

Cherry-picked from: f9fc942757

(cherry picked from commit e5b942ae007d0533fbd599c64d550f3a8355b940)
---
 doc/ChangeLog     | 5 +++++
 src/auths/plaintext.c | 6 +-----
 2 files changed, 6 insertions(+), 5 deletions(-)

diff --git a/doc/ChangeLog b/doc/ChangeLog
index 6492d2d1..b63226b9 100644
--- a/doc/ChangeLog
+++ b/doc/ChangeLog
@@ -40,6 +40,11 @@ JH/10 OpenSSL: Fix aggregation of messages.  Previously, when PIPELINING was
       dropped connections and sometimes bounces generated by a peer sending
       to this system.
 
+JH/11 Harden plaintext authenticator against a badly misconfigured client-send
+      string.  Previously it was possible to cause undefined behaviour in a
+      library routine (usually a crash).  Found by "zerons".
+
+
 
 Exim version 4.92
 -----------------
diff --git a/src/auths/plaintext.c b/src/auths/plaintext.c
index 7a0f7885..fa05b0ad 100644
--- a/src/auths/plaintext.c
+++ b/src/auths/plaintext.c
@@ -223,11 +223,7 @@ while ((s = string_nextinlist(&text, &sep, big_buffer, big_buffer_size)))
       if (ss[i+1] != '^')
 	ss[i] = 0;
       else
-        {
-        i++;
-        len--;
-        memmove(ss + i, ss + i + 1, len - i);
-        }
+        if (--len > ++i) memmove(ss + i, ss + i + 1, len - i);
 
   /* The first string is attached to the AUTH command; others are sent
   unembellished. */
-- 
2.20.1

